### ISSUE TITLE
Version 0.3.3 doesn't work - worker not connecting

### ISSUE THREAD

**Issue by KillyMXI** [Gwenio/vscode-ava-test-adapter#11]
(Following directions from https://github.com/Gwenio/vscode-ava-test-adapter/issues/6#issuecomment-536044953.)

```
> node --version
v10.16.3
```
VSCode About:
> Version: 1.38.1 (user setup)
> Commit: b37e54c98e1a74ba89e03073e5a3761284e3ffb0
> Date: 2019-09-11T13:35:15.005Z
> Electron: 4.2.10
> Chrome: 69.0.3497.128
> Node.js: 10.11.0
> V8: 6.9.427.31-electron.0
> OS: Windows_NT x64 10.0.18362

Test Explorer shows infinite reloading animation. Not even root nodes shown in the tests tree.

AVA Tests output empty.

AVA Explorer Log:
```
[2019-09-26 20:27:35.299] [INFO] Test Explorer found
[2019-09-26 20:27:35.864] [INFO] Creating adapter for <project folder>
[2019-09-26 20:27:35.864] [INFO] Initializing AVA Adapter...
[2019-09-26 20:27:35.865] [INFO] Using default NodeJS: C:\nodejs\node.exe
[2019-09-26 20:27:35.866] [DEBUG] Spawning worker...
[2019-09-26 20:27:35.871] [INFO] Registering adapter for <project folder>
[2019-09-26 20:27:35.871] [INFO] Loading test files of <project folder>
[2019-09-26 20:27:35.871] [INFO] Initialization finished
```

Nothing related to this extension is reported in the Dev Tools console either.

I checked for anything that can be associated with this extension in the VSCode process explorer but found nothing.
After issuing `Developer: Reload Window` command, few things flicked for a moment in the process explorer, leaving no time to identify them, but nothing stayed longer that that to vanish later.

**Comment by Gwenio** [comment]
For the next release the main goal for this issue will be to ensure the intended timeout occurs to prevent hanging.

If it does not hang, then a new attempt to spawn and connect to the worker can be forced by changing some of the settings (cwd, env, nodePath, nodeArgv).

There will be some clean up of the spawn/connect code,included, but it is hard to say if it will fix the issue.

Also, the time to wait before timing out the connection will be made configurable. This will make it more convienant to add '--inspect-brk=<port #>' to nodeArgv and connect the debugger to the worker process.

The release should be ready on Thursday.

**Comment by Gwenio** [comment]
@KillyMXI  New version is out.
It should at least ensure the extension does not hang forever waiting to connect to the worker.

If it does not get hung but is still not working, it is now possible to debug the worker process from VSCode.
How to do so is documented in the README file.
This will allow for getting a better idea of what is going wrong.

**Comment by KillyMXI** [comment]
After the update, I didn't have the infinite reloading animation anymore.

AVA Explorer Log had following:
```
[2019-10-05 20:37:21.619] [INFO] Test Explorer found
[2019-10-05 20:37:21.960] [INFO] Creating adapter for <project folder>
[2019-10-05 20:37:21.961] [INFO] Initializing AVA Adapter...
[2019-10-05 20:37:21.961] [INFO] Using default NodeJS: C:\nodejs\node.exe
[2019-10-05 20:37:21.961] [ERROR] Found invalid value for avaExplorer.nodePath { [ArgumentError: Any predicate failed with the following errors:
- Expected `nodePath` to be of type `undefined` but received type `string`
- Expected string `nodePath` to not be empty
	at Object.e.exports.Object.defineProperties.create.value.r [as nodePath] (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:21216)
	at U.getNode (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:56750)
	at U.setValue (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:58160)
	at new U (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:56373)
	at new me (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:69960)
	at r.TestAdapterRegistrar.exports.activate.e.subscriptions.push.e [as adapterFactory] (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:75205)
	at r.TestAdapterRegistrar.add (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:1594)
	at new r.TestAdapterRegistrar (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:1094)
	at exports.activate (c:\Users\Killy\.vscode\extensions\gwenio.vscode-ava-test-adapter-0.4.0\dist\main.js:18:75194)] name: 'ArgumentError' }
[2019-10-05 20:37:21.963] [DEBUG] Spawning worker...
[2019-10-05 20:37:21.971] [INFO] Registering adapter for <project folder>
[2019-10-05 20:37:21.971] [INFO] Loading test files of <project folder>
[2019-10-05 20:37:21.971] [INFO] Initialization finished
[2019-10-05 20:37:27.885] [DEBUG] Worker connected.
[2019-10-05 20:37:27.895] [INFO] Loaded test information for ava.config.js
```

I didn't have "avaExplorer.nodePath" set to anything. Why it became an empty string?

I tried to set it to `null` or `"null"` to see whether it will be able to use the built in Node. Both gave me different errors.

I then set it to working path:
```
"avaExplorer.nodePath": "C:\\nodejs\\node.exe"
```
Log showed that everything must be fine now:
```
[2019-10-05 20:55:36.225] [INFO] Test Explorer found
[2019-10-05 20:55:36.527] [INFO] Creating adapter for <project folder>
[2019-10-05 20:55:36.527] [INFO] Initializing AVA Adapter...
[2019-10-05 20:55:36.527] [INFO] Using default NodeJS: C:\nodejs\node.exe
[2019-10-05 20:55:36.528] [DEBUG] Spawning worker...
[2019-10-05 20:55:36.534] [INFO] Registering adapter for <project folder>
[2019-10-05 20:55:36.535] [INFO] Loading test files of <project folder>
[2019-10-05 20:55:36.535] [INFO] Initialization finished
[2019-10-05 20:55:37.930] [DEBUG] Worker connected.
[2019-10-05 20:55:37.937] [INFO] Loaded test information for ava.config.js
```
But! No actual tests were added to the tree.

After I clicked on "Reload Tests" button in the Test Explorer, they were finally added:
```
[2019-10-05 20:57:14.742] [INFO] Loading test files of <project folder>
[2019-10-05 20:57:20.774] [INFO] c65212a9d is the ID of config <project folder>\ava.config.js
[2019-10-05 20:57:20.774] [INFO] Received test file prefix <project folder>\tests\ from worker
[2019-10-05 20:57:20.776] [INFO] Received test file f409caf86 from worker
[2019-10-05 20:57:20.776] [INFO] Received test case t6a4f3e7f48 of file f409caf86 from worker
[2019-10-05 20:57:20.776] [INFO] Received test case t0bc3555044 of file f409caf86 from worker
[2019-10-05 20:57:20.776] [INFO] Received test case tca018c3c41 of file f409caf86 from worker
[2019-10-05 20:57:20.776] [INFO] Received test case t1e94ea5c4a of file f409caf86 from worker
...
[2019-10-05 20:57:20.780] [INFO] Received test file f39124ff8 from worker
[2019-10-05 20:57:20.780] [INFO] Received test case t113d250838 of file f39124ff8 from worker
[2019-10-05 20:57:20.780] [INFO] Received test case t91874fd040 of file f39124ff8 from worker
[2019-10-05 20:57:20.780] [INFO] Loaded test information for ava.config.js
```

I guess the worker issue is fixed now.
Is there anything non-trivial about nodePath and tests not being loaded automatically to check and open separate issues?

**Comment by Gwenio** [comment]
@KillyMXI 
> I guess the worker issue is fixed now.

Closing this issue them.

> Is there anything non-trivial about nodePath and tests not being loaded automatically to check and open separate issues?

I made a [issue for that](https://github.com/Gwenio/vscode-ava-test-adapter/issues/13). More specifically, the tests for the first AVA configuration not getting enumerated.
The IPC connection is being used before it is completely ready, leading to messages getting lost.

I would have got back about this sooner, but have been very busy preparing the code base to create a wrapper for AVA's internals that the extension uses.
The wrapper will be an NPM package that will need to be installed in the the project, and will keep changes in AVA from breaking the extension. Provided the appropriate version of the wrapper is installed, of course.
